src/charm.py

```diff
@@ -80,11 +86,50 @@ def _on_config_changed(self, _):
         else:
             self.unit.status = WaitingStatus("Waiting for Pebble in workload container")
 
+    def _patch_auth_resources(self) -> bool:
+        """Returns True if auth resources are patched correctly, False otherwise"""
+        self.unit.status = MaintenanceStatus("Patching auth configuration")
+        client = Client()
+        context = {"namespace": self._namespace, "app_name": self._name}
+
+        with open(Path(self._src_dir) / "auth_manifests.yaml") as f:
+            for obj in codecs.load_all_yaml(f, context=context):
+                try:
+                    client.patch(
+                        type(obj), obj.metadata.name, obj, patch_type=PatchType.MERGE
+                    )
+                except ApiError as e:
+                    logging.error(f"Patching auth resources failed: {str(e)}")
+                    return False
+        logging.info("Auth resources successfully patched.")
+        return True
+
+    def _create_crds(self) -> bool:
+        """Returns True if CRDs are created correctly, False otherwise."""
+        self.unit.status = MaintenanceStatus("Creating CRDs")
+        client = Client()
+
+        with open(Path(self._src_dir) / "crds_manifests.yaml") as f:
+            for obj in codecs.load_all_yaml(f):
```

Comment:
It really seems like we should be propagating the errors and letting the event handler catch them, for something like this:

```python
try:
    client.create(obj)
except ApiError as e:
    if e.status.reason == 'AlreadyExists:
        # ignore
    else:
        raise
```
and
```python
  try:
      ...
  except ApiError as err:
      logging.error("Applying resources failed.")
      logging.error(traceback.format_exc())
      self.unit.status = BlockedStatus(str(err))
```

Returning a `bool` here doesn't tell the caller anything about why it failed, which means the `_on_install` handler can't set a useful `BlockedStatus` message.

Reply:
As of now, `_on_install` doesn't handle the case where either `_create_crds()` or `_patch_auth` return `False`, but that's something I'll add once we decide how to correctly tell users what to do in the event of failure. I am thinking on something like:
`_create_crds():`

```python
  try:
  ...
  except ApiError as err:
      logging.error("<specific resource> failed.") # Let users know exactly which resource it tried to patch/create
      logging.error(err.<data>) # Exception data
      return False
```
`_on_install():`

```python
if True:
  self.unit.status = ActiveStatus()
else:
  self.unit.status = BlockedStatus("Resources failed, please do <action>"))
```

As someone commented previously, logging is okay in helper methods, but setting the status should happen in the event handler. Does this make sense?

Reply:
Agree on the latter point, but I think what @knkski is saying is, you could:

```python
try:
    helper_methods
except ApiError as e:
    logging.error("something")
    self.unit.status = BlockedStatus("helllllllp")
else:
    self.unit.status = ActiveStatus()
```

Reply:
Oh, I get it. Then +1.